cfpb/capital-framework
# Add cf-modals

Features:

Modal can be closed with:
 - [x] A delete minicon or label in the upper right corner
 - [x] The next action button or link
 - [x] Anywhere outside the modal window
 - [x] ESC key on a keyboard

Accessibility:
- [x] The close minicon needs to be accessible from a keyboard.
- [x] The enter key should be mapped to any submission forms in the modal (hit enter while in form input).
- [x] Provide separate focus and hover states for the close minicon and any “next” action buttons.
- [x] Modal windows should be marked with aria-hidden="true" and toggled to false when visible.
- [x] Modals should be given the role=dialog ARIA role.
- [x] Add role="alertdialog" to the modal window requiring user interaction.
- [x] Code the close minicon using a button element for more semantic markup.
- [x] The heading of the modal dialog should be marked as an h1.

## Additions

- Adds cf-modals 1.0.

## Testing

- Check out this branch.
- Run `npm run cf-link`
- Change to another clone of capital-framework.
- Run `git fetch && git checkout gh-pages-modals`
- Run `npm link cf-modals`
- Run `npm run build && npm start`
- Check http://localhost:3000/components/cf-modals/
- Edit the page source to add `o-modal__visible` to the modal you want to show and reload the page.
- Inspect the page and see the aria- attributes toggling when hiding the modal.

## Todos:

These are taken from http://cfpb.github.io/design-manual/page-components/modals.html

- The find function (ctrl+F) will not search information contained within a modal window. _I don't think this is possible?_
- Include offscreen instructions that describe the modal dialog and how to interact with it. Offscreen interaction instructions should use aria-labelledby attribute. _I didn't find any information about this on https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alertdialog_role_
- When the modal dialog is displayed, the main content of the page should be marked with aria-hidden="true" to prevent screen readers from interacting with it.
- The tab order inside the alert dialog must wrap. 

## Screenshots

Dialog modal:
![screen shot 2017-09-12 at 1 21 43 am](https://user-images.githubusercontent.com/704760/30309339-de08dd58-9758-11e7-9780-489e33e20d19.png)

Alert modal:
![screen shot 2017-09-12 at 1 21 26 am](https://user-images.githubusercontent.com/704760/30309342-e3f54684-9758-11e7-90bd-4d226fb64dbf.png)

## Notes

Modals requires some new additions to how CF works in regard to JS. Modals are absolutely positioned on the page, so can't be inside a container that has anything other than `position: static` set, otherwise it will end up relative to the container. This very likely means that there needs to be two components for modals: the actual modal and a "ModalTrigger" molecule (likely a link or button), which live in different areas of the page's code. The problem is we don't have a mechanism in place to communicate between the JS of two components on the page, which would be needed for a ModalTrigger to tell a Modal's JS to add the `o-modal__visible` class. @sebworks suggested that a redux utility be added to AtomicComponent and that @contolini had done work on this in cfgov-refresh already. This PR only provides the code structure for Modals, and were it to be visible, the logic for it to be closed.
